### PR TITLE
Adding py_gnome and oil_library

### PR DESCRIPTION
**Pull request by ocefpaf**
See https://github.com/NOAA-ORR-ERD/PyGnome/issues/12#issuecomment-170897658

### CODE REVIEW

**Comment by ocefpaf** [comment]
@ChrisBarker-NOAA I am trying to build `py_gnome` and `oil_library`.  And I have a few questions for you.

For `oil_library` do we need the `remake_oil_db` step?

Now... For `py_gnome`...

1) I can `python setup.py install` using conda and following the instructions from the [site](https://github.com/NOAA-ORR-ERD/PyGnome/blob/master/InstallingWithAnaconda.rst), but then I cannot import the module:
```python
python gnome
"ImportError: libgnome.so: cannot open shared object file: No such file or directory
```

@daf's [package](https://anaconda.org/DAF/pygnome) suffer from the same issue BTW.

The file `libgnome.so` does exist, so...

2) I tried to debug this by building the cython module with `python  setup.py build` and that got me a 

    ld: cannot find -lgnome

Here is the full error message.

```shell
running build_ext
skipping 'gnome/cy_gnome/cy_basic_types.cpp' Cython extension (up-to-date)
building 'gnome.cy_gnome.cy_basic_types' extension
creating build/temp.linux-x86_64-2.7/gnome
creating build/temp.linux-x86_64-2.7/gnome/cy_gnome
gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -DpyGNOME=1 -I../lib_gnome -I/home/filipe/miniconda/envs/_test/lib/python2.7/site-packages/numpy/core/include -I../third_party_lib/netcdf-4.3/linux2/x86_64/include -I. -I/home/filipe/miniconda/envs/_test/include/python2.7 -c gnome/cy_gnome/cy_basic_types.cpp -o build/temp.linux-x86_64-2.7/gnome/cy_gnome/cy_basic_types.o
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++ [enabled by default]
g++ -pthread -shared -L/home/filipe/miniconda/envs/_test/lib -Wl,-rpath=/home/filipe/miniconda/envs/_test/lib,--no-as-needed -L/home/filipe/miniconda/conda-bld/work/py_gnome/gnome/cy_gnome -Wl,-rpath=$ORIGIN build/temp.linux-x86_64-2.7/gnome/cy_gnome/cy_basic_types.o -L/home/filipe/miniconda/envs/_test/lib -lgnome -lpython2.7 -o build/lib.linux-x86_64-2.7/gnome/cy_gnome/cy_basic_types.so
/usr/lib64/gcc/x86_64-suse-linux/4.8/../../../../x86_64-suse-linux/bin/ld: cannot find -lgnome
```

The worst part is that I cannot build even the broken version of the package using our CentOS6 docker image. (See the [full log](https://circleci.com/gh/ioos/conda-recipes/883) for details.)

And we do have to use the CentOS6 image to make the package compatible with all Linux (GLBIC)  versions.

Any advice?

**Comment by rsignell-usgs** [comment]
@ocefpaf , the build errors seem to indicate that the netcdf library and include files are not being found:
```
../lib_gnome/TimeGridVel_c.cpp:6168: error: ‘NC_NOWRITE’ was not declared in this scope
../lib_gnome/TimeGridVel_c.cpp:6168: error: ‘nc_open’ was not declared in this scope
../lib_gnome/TimeGridVel_c.cpp:6169: error: ‘NC_NOERR’ was not declared in this scope
../lib_gnome/TimeGridVel_c.cpp:6171: error: ‘nc_inq_ndims’ was not declared in this scope
../lib_gnome/TimeGridVel_c.cpp:6172: error: ‘NC_NOERR’ was not declared in this scope
../lib_gnome/TimeGridVel_c.cpp:6191: error: ‘nc_inq_varid’ was not declared in this scope
```

**Comment by ChrisBarker-NOAA** [comment]
> For oil_library do we need the remake_oil_db step?

yes -- there is a CSV file that should be in the repo -- and that step builds a sqlite DB from that.

We mostly run in "develop" mode, and I think there is an issue with the DB file getting installed if you try to install it. I'll take a look at that -- it's on my list anyway.

**Comment by ChrisBarker-NOAA** [comment]
python gnome
"ImportError: libgnome.so: cannot open shared object file: No such file or directory
@daf's package suffer from the same issue BTW.

interesting -- I know he got it working one way or another...

But I think Rich is right -- this is likely the netcdf libs -- lib_gnome (all our C++ code) needs those, in addition to the python netcdf package needing them.

**Comment by ChrisBarker-NOAA** [comment]
on CentOS -- do you use rpms for netcdf, or conda-provided versions?

so far, we haven't done conda builds on Linux... though that's on my list, too :-)

**Comment by ocefpaf** [comment]
> yes -- there is a CSV file that should be in the repo -- and that step builds a sqlite DB from that.

I asked because I think the `install` command is doing that, but I saw somewhere that added as a `post-link.sh` step.

> But I think Rich is right -- this is likely the netcdf libs -- lib_gnome (all our C++ code) needs those, in addition to the python netcdf package needing them.

Those are installed and that error only happens on the CentOS6 machine.  I can build the broken package locally without that error.

> on CentOS -- do you use rpms for netcdf, or conda-provided versions?

The libs are from conda's default channel.

> interesting -- I know he got it working one way or another...

I cannot install that and import, but I able to enter the conda work directory and import from there.
Maybe there is a packaging issue somewhere...

**Comment by ocefpaf** [comment]
> we haven't done conda builds on Linux... though that's on my list, too :-)

The CIs results confirm that :wink: (Only the OSX build returned green.)

**Comment by ocefpaf** [comment]
@daf any advice?  Can you confirm that your binary works on your setup?

**Comment by ocefpaf** [comment]
Digging into this I found out that:

```shell
> ldd cy_basic_types.so                                                                                                                                                  ⏎
        linux-vdso.so.1 (0x00007ffd099ea000)
        libgnome.so => not found
        libpython2.7.so.1.0 => /home/filipe/miniconda/envs/_test/lib/python2.7/site-packages/gnome/cy_gnome/./../../../../libpython2.7.so.1.0 (0x00007f2229806000)
        libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f22294fe000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f22291fd000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f2228fe5000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f2228dc8000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f2228a21000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f222881c000)
        libutil.so.1 => /lib64/libutil.so.1 (0x00007f2228619000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f2229e1e000)
```

This can be worked around by setting `LD_LIBRARY_PATH`, but that is not recommended.

    export LD_LIBRARY_PATH=$HOME/miniconda/envs/_test/lib/python2.7/site-packages/gnome/cy_gnome:$LD_LIBRARY_PATH

**Comment by ocefpaf** [comment]
The AppVeyor failure due to some sort of corruption of the binary for `msinttypes`.  The same package exists in scitools and conda-forge and they are OK.  I will delete ours and re-add it.

**Comment by ocefpaf** [comment]
@ChrisBarker-NOAA there are lots of things that makes this hard to fix/debug.

- The netCDF problem on OSX is solved because the repository brings pre-complied versions of the library! (See https://github.com/NOAA-ORR-ERD/PyGnome/blob/0ecb501da7e16895fafeee6f62c0202047eaa505/py_gnome/setup.py#L208-L215)
- To get this working on Linux we need to patch that `setup.py` [here](https://github.com/NOAA-ORR-ERD/PyGnome/blob/0ecb501da7e16895fafeee6f62c0202047eaa505/py_gnome/setup.py#L348-L352).
- The `if`-clauses define something for `darwin` or `win32` (and one time for `linux2`), but do nothing for `else`.  That means some variables will end up undefined under certain circumstances.
- There are many parts that are kept in the root directory, like the extensions, and that makes it hard to follow what is going on. (See https://github.com/NOAA-ORR-ERD/PyGnome/blob/0ecb501da7e16895fafeee6f62c0202047eaa505/py_gnome/setup.py#L330-L331)
Ideally those should be in their own separate package.  (Not sure if that makes sense for `py_gnome` though.)
- I see that `setup.py` checks for `sys.platform == linux2`.  Maybe it is safer look for `startswith('linux')`
- In https://github.com/NOAA-ORR-ERD/PyGnome/blob/0ecb501da7e16895fafeee6f62c0202047eaa505/py_gnome/setup.py#L435 the Linux `include_dirs` seems to be ill defined.  (Missing `np.get_include()`? And maybe something to make it find conda's libnetcdf too.)
- Last, but not least, I am not sure about this line: https://github.com/NOAA-ORR-ERD/PyGnome/blob/0ecb501da7e16895fafeee6f62c0202047eaa505/py_gnome/setup.py#L460. It seems to work on OSX though.

Let me know if you have some idea where to start.  (My approach to fix this would not be very "time effective.")

**Comment by rsignell-usgs** [comment]
@ocefpaf, jeez, thanks for all the hard work here.  I feel a bit guilty for getting you into this, but I'm sure  @ChrisBarker-NOAA and Co appreciate!

**Comment by ChrisBarker-NOAA** [comment]
Thanks!

Agreed -- this is best cleaned up by us. It's built up a lot of kruft over time, and we got it working with all our developers, but it really should be cleaned up to be packaged better.

Though we've got it all building on Linux, so I'm not sure what's up there. But on my list to get it working with conda on Linux, so that'll motivate.

**Comment by ocefpaf** [comment]
> Though we've got it all building on Linux, so I'm not sure what's up there. 

I can get it to work by setting `LD_LIBRARY_PATH`.  That is OK while developing, but that is not recommend when using packaging with conda.

**Comment by kwilcox** [comment]
I believe @daf has the entire pygnone system packaged in conda (including oillibrary and the web ui)

**Comment by ChrisBarker-NOAA** [comment]
On Jan 12, 2016, at 3:51 PM, Kyle Wilcox <notifications@github.com> wrote:

I believe @daf <https://github.com/daf> has the entire pygnone system
packaged in conda (including oillibrary and the web ui)

Cool! @daf--if you needed to do anything special, please let us know.

-Chris


—
Reply to this email directly or view it on GitHub
<https://github.com/ioos/conda-recipes/pull/699#issuecomment-171102987>.

**Comment by ocefpaf** [comment]
TL;DR &rarr; https://github.com/ioos/conda-recipes/pull/699#issuecomment-170998961

Thanks @kwilcox.  This recipe here **is** based on @daf's package that @rsignell-usgs found [here](https://anaconda.org/DAF/pygnome).

That version of his package is not installing a few dependencies at run time, like `unit_conversion`, making it hard for people to just `conda -c install pygnome` since they are left out to figure the rest.

Also, that package does seem to suffer from the same issue reported [here](https://github.com/ioos/conda-recipes/pull/699#issuecomment-170984225).  Confirm in my setup and @rsignell-usgs'.

In addition, that package seems to have been be compiled locally and not using a clean CI.  That means a few undocumented extra libraries might be needed. BTW  I can compile it locally with my libnetcdf, but then it won't work with conda's libnetcdf. (See https://circleci.com/gh/ioos/conda-recipes/883).

**Comment by ocefpaf** [comment]
@rsignell-usgs now that I fixed https://github.com/ioos/conda-recipes/pull/699#issuecomment-171044904 the AppVeyor builds are working.  The trick is the same as OSX: bundled libnetCDF. I will see if I can work my way around the setup.py and patch it to use conda's libnetcdf on Linux.

That should get us a hacky, but working, package.

**Comment by rsignell-usgs** [comment]
So go ahead and accept this PR?

**Comment by ocefpaf** [comment]
If you do not need Linux I can skip the build and yes.  If you do give me some more time.  (I am looking into the problem.)

**Comment by ocefpaf** [comment]
@rsignell-usgs the binaries for Windows and OSX will be in the channel soon.  Let me know how it goes if test them.

I opened a new issue for to complete the Linux package. (See #703.)

**Comment by daf** [comment]
I didn't have to do anything special to get it linked to that .so!  I'll look at my setup and see if I can find anything that looks pertinent.

(github, would've been nice to get emails from this thread since I'm mentioned..)

**Comment by ocefpaf** [comment]
@daf take a look at this issue I just opened: https://github.com/NOAA-ORR-ERD/PyGnome/issues/15

**Comment by daf** [comment]
See https://github.com/NOAA-ORR-ERD/PyGnome/issues/12#issuecomment-171303745

**Comment by ChrisBarker-NOAA** [comment]
NOTE: py_gnome and the oil_library are separate python packages, because theoretically, at some point in the future we may use the oil_libary by itself. But in practice, they are being developed very closely in concert, hence why they are in the same repo.

So it may make sense to simply have a single conda package for both of them.

**Comment by ocefpaf** [comment]
> So it may make sense to simply have a single conda package for both of them.

It is easier to manage a package for each at this point.

**Comment by ChrisBarker-NOAA** [comment]
On Wed, Jan 13, 2016 at 9:47 AM, Filipe <notifications@github.com> wrote:

> So it may make sense to simply have a single conda package for both of
> them.
>
> It is easier to manage a package for each at this point.
>
> OK -- brings up a good point -- we should be able to run py_gnome without
the oil_library -- if you dont want to model oil weathering. I should
probably make sure that's possible.

and vice-versa....




> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ioos/conda-recipes/pull/699#issuecomment-171377217>.
>



-- 

Christopher Barker, Ph.D.
Oceanographer

Emergency Response Division
NOAA/NOS/OR&R            (206) 526-6959   voice
7600 Sand Point Way NE   (206) 526-6329   fax
Seattle, WA  98115       (206) 526-6317   main reception

Chris.Barker@noaa.gov
